lib/sycamore/sycamore/llms/openai.py

```diff
@@ -295,61 +338,97 @@ def _get_cache_key(self, prompt_kwargs: dict, llm_kwargs: Optional[dict] = None)
+            prompt_kwargs={prompt_kwargs} 
+            llm_kwargs={llm_kwargs} 
+            model_name={self.model.name}"""
+            return (key, hit.get("result"))
+        return (key, None)
+
+    def _cache_set(self, key, result):
+        if key is None:
+            return
+        self._cache.set(key, result)
+
+    def _get_generate_kwargs(self, prompt_kwargs: dict, llm_kwargs: Optional[dict] = None) -> dict:
         kwargs = {
             "temperature": 0,
-            **llm_kwargs,
+            **(llm_kwargs or {}),
         }
-
         if "SYCAMORE_OPENAI_USER" in os.environ:
             kwargs.update({"user": os.environ.get("SYCAMORE_OPENAI_USER")})
 
         if "prompt" in prompt_kwargs:
             prompt = prompt_kwargs.get("prompt")
-            messages: list[ChatCompletionMessageParam] = [{"role": "user", "content": f"{prompt}"}]
+            kwargs.update({"messages": [{"role": "user", "content": f"{prompt}"}]})
         elif "messages" in prompt_kwargs:
-            messages = prompt_kwargs["messages"]
+            kwargs.update({"messages": prompt_kwargs["messages"]})
         else:
             raise ValueError("Either prompt or messages must be present in prompt_kwargs.")
+        return kwargs
+
+    def generate(self, *, prompt_kwargs: dict, llm_kwargs: Optional[dict] = None) -> str:
+        key, ret = self._cache_get(prompt_kwargs, llm_kwargs)
+        if ret is not None:
+            return ret
+
+        if llm_kwargs is not None:
+            ret = self._generate_using_openai(prompt_kwargs, llm_kwargs)
+        else:
+            ret = self._generate_using_guidance(prompt_kwargs)
+
+        value = {
+            "result": ret,
+            "prompt_kwargs": prompt_kwargs,
+            "llm_kwargs": llm_kwargs,
+            "model_name": self.model.name,
+        }
+        self._cache_set(key, value)
+        return ret
+
+    def _generate_using_openai(self, prompt_kwargs, llm_kwargs) -> str:
+        kwargs = self._get_generate_kwargs(prompt_kwargs, llm_kwargs)
+
+        completion = self.client_wrapper.get_client().chat.completions.create(model=self._model_name, **kwargs)
+        return completion.choices[0].message.content
+
+    async def generate_async(self, *, prompt_kwargs: dict, llm_kwargs: Optional[dict] = None) -> Awaitable[str]:
+        key, ret = self._cache_get(prompt_kwargs, llm_kwargs)
+        if ret is not None:
+            return ret
+
+        if llm_kwargs is None:
+            raise ValueError("Must include llm_kwargs to generate future call")
+        ret = await self._generate_future_using_openai(prompt_kwargs, llm_kwargs)
+
+        value = {
+            "result": ret,
+            "prompt_kwargs": prompt_kwargs,
+            "llm_kwargs": llm_kwargs,
+            "model_name": self.model.name,
+        }
+        self._cache_set(key, value)
+        return ret
+
+    async def _generate_future_using_openai(self, prompt_kwargs, llm_kwargs) -> Awaitable[str]:
```

Comment:
I'd rename this to _generate_awaitable_using_openai since it isn't strictly making a future.

Reply:
oops! sorry missed that. changed.